pin/set.go

```diff
@@ -0,0 +1,337 @@
+package pin
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"sort"
+	"unsafe"
+
+	"github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/gogo/protobuf/proto"
+	"github.com/ipfs/go-ipfs/merkledag"
+	"github.com/ipfs/go-ipfs/pin/internal/pb"
+	"github.com/ipfs/go-ipfs/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultFanout = 256
+	maxItems      = 8192
+)
+
+func randomSeed() (uint32, error) {
+	var buf [4]byte
+	if _, err := rand.Read(buf[:]); err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint32(buf[:]), nil
+}
+
+func hash(seed uint32, k util.Key) uint32 {
+	var buf [4]byte
+	binary.LittleEndian.PutUint32(buf[:], seed)
+	h := fnv.New32a()
+	_, _ = h.Write(buf[:])
+	_, _ = io.WriteString(h, string(k))
```

Comment:
http://golang.org/pkg/hash/#Hash "It never returns an error.". I use the _ there to remind myself of that.